gluten-core/src/main/scala/org/apache/spark/memory/SparkMemoryUtil.scala

```diff
@@ -209,9 +209,50 @@ object SparkMemoryUtil {
     sb.toString()
   }
 
+  def addLeakSafeTaskCompletionListener[U](f: TaskContext => U): TaskContext = {
+    TaskContext.get().addTaskCompletionListener(f)
+  }
+
   private case class PrintableMemoryUsageStats(
       name: String,
       used: Option[Long],
       peak: Option[Long],
       children: Iterable[PrintableMemoryUsageStats])
+
+  class GenericRetainer[T <: AutoCloseable] {
+    private var retained: Option[T] = None
+
+    def retain(batch: T): Unit = {
+      if (retained.isDefined) {
+        throw new IllegalStateException
+      }
+      retained = Some(batch)
+    }
+
+    def release(): Unit = {
+      retained.foreach(b => b.close())
+      retained = None
+    }
+  }
+
+  class UnsafeItr[T <: AutoCloseable](delegate: Iterator[T]) extends Iterator[T] {
+    val holder = new GenericRetainer[T]()
+
+    addLeakSafeTaskCompletionListener[Unit](
+      (_: TaskContext) => {
+        holder.release()
+      })
+
+    override def hasNext: Boolean = {
+      holder.release()
+      val hasNext = delegate.hasNext
+      hasNext
+    }
+
+    override def next(): T = {
+      val b = delegate.next()
+      holder.retain(b)
+      b
+    }
+  }
```

Comment:
Had some new APIs to replace this sort of wrappers

https://github.com/apache/incubator-gluten/blob/main/gluten-core/src/main/scala/org/apache/gluten/utils/Iterators.scala